backends/gdx-backends-gwt/src/com/badlogic/gdx/backends/gwt/GwtInput.java

```diff
@@ -1,1087 +1,1091 @@
+import com.badlogic.gdx.utils.TimeUtils;
+import com.google.gwt.core.client.JavaScriptObject;
+import com.google.gwt.core.client.JsArray;
+import com.google.gwt.dom.client.CanvasElement;
+import com.google.gwt.dom.client.Document;
+import com.google.gwt.dom.client.NativeEvent;
+import com.google.gwt.dom.client.Touch;
+import com.google.gwt.event.dom.client.KeyCodes;
+
+public class GwtInput implements Input {
+	static final int MAX_TOUCHES = 20;
+	boolean justTouched = false;
+	private IntMap<Integer> touchMap = new IntMap<Integer>(20);
+	private boolean[] touched = new boolean[MAX_TOUCHES];
+	private int[] touchX = new int[MAX_TOUCHES];
+	private int[] touchY = new int[MAX_TOUCHES];
+	private int[] deltaX = new int[MAX_TOUCHES];
+	private int[] deltaY = new int[MAX_TOUCHES];
+	IntSet pressedButtons = new IntSet();
+	int pressedKeyCount = 0;
+	IntSet pressedKeySet = new IntSet();
+	boolean[] pressedKeys = new boolean[256];
+	boolean keyJustPressed = false;
+	boolean[] justPressedKeys = new boolean[256];
+	boolean[] justPressedButtons = new boolean[5];
+	InputProcessor processor;
+	long currentEventTimeStamp;
+	final CanvasElement canvas;
+	final GwtApplicationConfiguration config;
+	boolean hasFocus = true;
+	GwtAccelerometer accelerometer;
+
+	public GwtInput (CanvasElement canvas, GwtApplicationConfiguration config) {
+		this.canvas = canvas;
+		this.config = config;
+		if (config.useAccelerometer) {
+			if (GwtApplication.agentInfo().isFirefox()) {
+				setupAccelerometer();
+			} else {
+				GwtPermissions.queryPermission(GwtAccelerometer.PERMISSION, new GwtPermissions.GwtPermissionResult() {
+					@Override
+					public void granted() {
+						setupAccelerometer();
+					}
+
+					@Override
+					public void denied() {
+					}
+
+					@Override
+					public void prompt() {
+						setupAccelerometer();
+					}
+				});
+			}
+		}
+		hookEvents();
+	}
+
+	void reset () {
+		if (justTouched) {
+			justTouched = false;
+			for (int i = 0; i < justPressedButtons.length; i++) {
+				justPressedButtons[i] = false;
+			}
+		}
+		if (keyJustPressed) {
+			keyJustPressed = false;
+			for (int i = 0; i < justPressedKeys.length; i++) {
+				justPressedKeys[i] = false;
+			}
+		}
+	}
+
+	void setupAccelerometer () {
+		if (GwtAccelerometer.isSupported()) {
+			if (accelerometer == null) accelerometer = GwtAccelerometer.getInstance();
+			if (!accelerometer.activated()) accelerometer.start();
+		}
+	}
+
+	@Override
+	public float getAccelerometerX () {
+		return this.accelerometer != null ? (float) this.accelerometer.x() : 0;
+	}
+
+	@Override
+	public float getAccelerometerY () {
+		return this.accelerometer != null ? (float) this.accelerometer.y() : 0;
+	}
+
+	@Override
+	public float getAccelerometerZ () {
+		return this.accelerometer != null ? (float) this.accelerometer.z() : 0;
+	}
+
+	private boolean isAccelerometerPresent() {
+		return getAccelerometerX() != 0 || getAccelerometerY() != 0 || getAccelerometerZ() != 0;
```

Comment:
Isn't there a better way to determine this? What about checking whether the accelerometer "activated" or "hasReading" returns true?

Reply:
It's indeed what is done here although it doesn't look like it. See GwtAccelerometer:

https://github.com/libgdx/libgdx/blob/24a6c9fb03eb7c8f8de17dcb71238a41a04cf4dc/backends/gdx-backends-gwt/src/com/badlogic/gdx/backends/gwt/GwtAccelerometer.java#L44

This line returns 0 exactly if the sensor data is not present (then it is null). There's no chance that a real accelerometer returns exactly 0, so we won't have false positives on this.

Reply:
I see, hmm, would be nice if it was more explicit but this is probably fine.